source/common/http/async_client_impl.h

```diff
@@ -466,25 +522,64 @@ class AsyncStreamImpl : public AsyncClient::Stream,
   friend class AsyncClientImplUnitTest;
 };
 
-class AsyncRequestImpl final : public AsyncClient::Request,
-                               AsyncStreamImpl,
-                               AsyncClient::StreamCallbacks {
+class AsyncRequestSharedImpl : public virtual AsyncClient::Request,
+                               protected AsyncStreamImpl,
+                               protected AsyncClient::StreamCallbacks {
 public:
-  AsyncRequestImpl(RequestMessagePtr&& request, AsyncClientImpl& parent,
-                   AsyncClient::Callbacks& callbacks, const AsyncClient::RequestOptions& options);
+  void cancel() final;
+
+protected:
+  AsyncRequestSharedImpl(AsyncClientImpl& parent, AsyncClient::Callbacks& callbacks,
+                         const AsyncClient::RequestOptions& options);
+  void onHeaders(ResponseHeaderMapPtr&& headers, bool end_stream) final;
+  void onData(Buffer::Instance& data, bool end_stream) final;
+  void onTrailers(ResponseTrailerMapPtr&& trailers) final;
+  void onComplete() final;
+  void onReset() final;
+
+  AsyncClient::Callbacks& callbacks_;
+  Tracing::SpanPtr child_span_;
+  std::unique_ptr<ResponseMessageImpl> response_;
+  bool cancelled_{};
+};
 
-  // AsyncClient::Request
-  void cancel() override;
+class AsyncOngoingRequestImpl final : public AsyncClient::OngoingRequest,
+                                      public AsyncRequestSharedImpl {
+public:
+  AsyncOngoingRequestImpl(RequestHeaderMapPtr&& request_headers, AsyncClientImpl& parent,
+                          AsyncClient::Callbacks& callbacks,
+                          const AsyncClient::RequestOptions& options)
+      : AsyncRequestSharedImpl(parent, callbacks, options),
+        request_headers_(std::move(request_headers)) {
+    ASSERT(request_headers_);
+  }
+  void captureAndSendData(Buffer::InstancePtr&& data, bool end_stream) override {
+    request_data_ = std::move(data);
```

Comment:
why move this to a class member variable?
if there's any expectations of persistence, another send may overwrite.

Reply:
Fixed.